internal/cli/server_upgrade.go

```diff
@@ -0,0 +1,362 @@
+				clierrors.Humanize(err),
+				terminal.WithErrorStyle(),
+			)
+			return 1
+		}
+		ctxName = defaultName
+	}
+
+	originalCfg, err := c.contextStorage.Load(ctxName)
+	if err != nil {
+		c.ui.Output(
+			"Error loading the context %q: %s",
+			ctxName,
+			clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+
+	// Upgrade waypoint server
+	sg := c.ui.StepGroup()
+	defer sg.Wait()
+
+	s := sg.Add("Validating server context: %q", ctxName)
+	defer func() { s.Abort() }()
+
+	conn, err := serverclient.Connect(ctx, serverclient.FromContextConfig(originalCfg))
+	if err != nil {
+		c.ui.Output(
+			"Error connecting with context %q: %s",
+			ctxName,
+			clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+
+	s.Update("Verifying connection is valid for context %q...", ctxName)
+
+	client := pb.NewWaypointClient(conn)
+	// validate API compat here with new clientpkg
+	if _, err := clientpkg.New(ctx,
+		clientpkg.WithLogger(c.Log),
+		clientpkg.WithClient(client),
+	); err != nil {
+		c.ui.Output(
+			"Error connecting with context %q: %s",
+			ctxName,
+			clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+
+	resp, err := client.GetVersionInfo(ctx, &empty.Empty{})
+	if err != nil {
+		c.ui.Output(
+			"Error retrieving server version info: %s", clierrors.Humanize(err),
+			terminal.WithErrorStyle())
+		return 1
+	}
+
+	initServerVersion := resp.Info.Version
+
+	s.Update("Context %q validated and connected successfully.", ctxName)
+	s.Done()
+
+	s = sg.Add("Starting server snapshots")
+
+	// Snapshot server before upgrade
+	if !c.skipSnapshot {
+		s.Update("Taking server snapshot before upgrading")
+
+		snapshotName := fmt.Sprintf("%s-%d", defaultSnapshotName, time.Now().Unix())
+		if c.snapshotName != "" {
+			snapshotName = fmt.Sprintf("%s-%d", c.snapshotName, time.Now().Unix())
+		}
+
+		s.Update("Taking snapshot of server with name: '%s'", snapshotName)
+		writer, err := os.Create(snapshotName)
+		if err != nil {
+			s.Update("Failed to take server snapshot")
+			s.Status(terminal.StatusError)
+			s.Done()
+
+			c.ui.Output(fmt.Sprintf("Error opening output: %s", err), terminal.WithErrorStyle())
+			return 1
+		}
+
+		err = clisnapshot.WriteSnapshot(c.Ctx, c.project.Client(), writer)
+		writer.Close()
+
+		if err != nil {
+			s.Update("Failed to take server snapshot")
+			s.Status(terminal.StatusError)
+			s.Done()
+
+			c.ui.Output(fmt.Sprintf("Error generating Snapshot: %s", err), terminal.WithErrorStyle())
+			return 1
+		}
+
+		s.Update("Snapshot of server written to: '%s'", snapshotName)
+		s.Done()
+	} else {
+		s.Update("Server snapshot disabled on request, this means no snapshot will be taken before upgrades")
+		s.Status(terminal.StatusWarn)
+		s.Done()
+		log.Warn("Server snapshot disabled on request from user, skipping")
+	}
+
+	c.ui.Output("Upgrading...", terminal.WithHeaderStyle())
+
+	c.ui.Output("Waypoint server will now upgrade from version %q",
```

Comment:
Oops, looks like re-arranging this removed the original PR review comment about this.

@mitchellh - So, I think the only way I can get this is by adding a `Config` to the Installer interface and exporting each platforms config? Not sure if that's a huge deal or if it's worth it for this message. Alternatively I could try to pass along the original version here to Upgrade or retrieve it again for each platform. Then at least there we'll have the version we're upgrading to easily.

Reply:
No, not a big deal today probably not worth the effort!